[FEATURE] Ajouter la méthode d'authentification GAR à l'utilisateur qui vient de se connecter à Mon-Pix (PIX-1095).

chrc opened the pull request:
## :unicorn: Problème
Lorsqu’un élève provenant du GAR tente de se réconcilier alors qu’**il possède déjà un compte (par e-mail ou identifiant) réconcilié** dans son établissement actuel ou dans un autre établissement, une erreur est renvoyée par l’API.
Cette erreur est ensuite affichée dans une modal qui l’invite à continuer en se connectant à son compte existant.
En cliquant sur “Poursuivre avec mon compte Pix”,  l'élève est redirigé vers la double mire de connexion afin qu’il puisse se connecter à son compte existant.
Mais la liste des méthodes d'authentification de ce compte n'est pas modifiée pour y ajouter la méthode GAR.

## :robot: Solution
Ajouter la méthode GAR au compte (par e-mail ou identifiant) avec lequel l'élève vient de se connecter depuis la double mire.

## :rainbow: Remarques
1 - Ce ticket implémente cette fonctionnalité lorsque que **l'élève est déjà réconcilié (avec e-mail ou identifiant)**.
<img src="https://user-images.githubusercontent.com/88607/92770616-65ffd680-f39a-11ea-89e4-0b375423eede.png">
<img src="https://user-images.githubusercontent.com/88607/92770644-7021d500-f39a-11ea-8939-4676122f539a.png" width="500">

2 - La fonctionnalité complémentaire (pas de réconciliation) sera implémentée dans un autre ticket.

3 - L'utilisateur doit être redirigé vers la campagne après s'être authentifié.
Pour cela, la méthode `this.session.attemptedTransition.retry()` a été surchargée afin de ne pas passer par la page par défaut (`/profile`).

## :100: Pour tester
### Se connecter via le GAR
* Avec pix-saml-idp
  * La configuration doit être modifiée pour être redirigé vers https://app-pr1841.review.pix.fr
  * Les variables d'environnement `SAML_IDP_CONFIG` et `SAML_SP_CONFIG` doivent être modifiées en conséquence
* Prénom = 'user', Nom = 'pix'
* Entrer le code campagne RESTRICTD
* Entrez la date de naissance 30/09/2010
* Se connecter en tant que user.pix@example.net / Pix123
* Vérifier qu'on arrive sur la campagne. 
* Se reconnecter via le GAR et vérifier qu'on ne passe plus par la réconciliation (et vérifier que samlID est renseigné).

chrc committed:
[API] Add method extractSamlId in tokenSenvice

chrc committed:
[API] Add method updateSamlId in user repository

chrc committed:
[API] Add usecase updateSamlId

chrc committed:
[API] Add updateUserSamlId in user controller

Commit:
[API] Create or update external user in usecase

Commit:
[API] Add acceptance test for external user reconcilation

chrc committed:
[APP] Add method to update GAR authenticationMethod

chrc committed:
[APP] Add method to add GAR authentication method

This controller method add the GAR authentication method to user,
with information inside the idToken.

chrc committed:
[APP] Add param @addGarAuthenticationMethodToUser

chrc committed:
[APP] Add GAR authentication method if needed

If IdToken exist add GAR authentication method to user

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr1841.review.pix.fr
- Orga: https://orga-pr1841.review.pix.fr
- Certif: https://certif-pr1841.review.pix.fr
- Admin: https://admin-pr1841.review.pix.fr
- API: https://api-pr1841.review.pix.fr/api/

 Please check it out!

chrc committed:
[APP] Add PATCH /users route into Mirage

yahyayouness commented on api/tests/unit/domain/services/token-service_test.js:
```diff
@@ -24,100 +27,94 @@ describe('Unit | Domain | Service | Token Service', function() {
+    it('should throw an InvalidTemporaryKeyError if the accessToken is invalid', async () => {
       // given
-      const token = 'eyJhbGciOiJIUzI1NiIsIgR5cCI6IkpXVCJ9.eyJ1c2VyX2lPIjoxMjMsImlhdCI6MTQ5OTA3Nzg2Mn0.FRAAoowTA8Bc6BOzD7wWh2viVN47VrPcGgLuHi_NmKw';
+      const accessToken = 'eyJhbGciOiJIUzI1NiIsIgR5cCI6IkpXVCJ9.eyJ1c2VyX2lPIjoxMjMsImlhdCI6MTQ5OTA3Nzg2Mn0.FRAAoowTA8Bc6BOzD7wWh2viVN47VrPcGgLuHi_NmKw';
```
C'est plutôt un id Token ici

chrc replied:
Done

yahyayouness commented on api/tests/unit/domain/services/token-service_test.js:
```diff
@@ -24,100 +27,94 @@ describe('Unit | Domain | Service | Token Service', function() {
       // when
-      const error = await catchErr(tokenService.extractExternalUserFromIdToken)(token);
+      const error = await catchErr(tokenService.extractExternalUserFromIdToken)(accessToken);
       expect(error).to.be.an.instanceof(InvalidTemporaryKeyError);
```
Je propose un BSR de renommer le InvalidTemporaryKeyError => InvalidTokenError en modifiant le message par défault : 'Invalid token' qui peut être override par un message plus préci à l'appel.

octo-topi replied:
token => accessToken corrigé

octo-topi replied:
Cette méthode `extractExternalUserFromIdToken`  est appelée uniquement par le use-case `create-user-and-reconcile-to-schooling-registration-from-external-user`, qui ne fait pas partie du scope du ticket. Comme vu ensemble, pas de BSR > 10 minutes sur ce ticket, ce sera fait dans un ticket dédié.

yahyayouness commented on api/lib/infrastructure/repositories/user-repository.js:
```diff
@@ -432,6 +433,19 @@ module.exports = {
       });
+  },
+
+  async updateSamlId({ userId, samlId }) {
```
Ca va faire doublon avec la methode que j'ai crée updateUserAttributes , qui met à jour tous attributs directs avec la même méthode.

chrc replied:
Dans l'absolue, en tant qu'_interface_ avec le `usecase` j'aurai préféré `addAuthenticateMethodGAR`.
Ensuite cette méthode pourrait faire appel à `updateUserAttributes`.

chrc replied:
+1 pour un BSR dans un autre ticket.

octo-topi replied:
Déjà échangé au récap tech et mentionné dans les standards: le fait de ne pas spécialiser le repository ou de lui faire porter un nom fonctionnel ne fait pas l'unanimité. Aucun argument définitif en faveur de l'une ou l'autres des solutions n'est connu, donc je propose de laisser tel quel.

yahyayouness commented on api/lib/domain/usecases/update-user-samlId.js:
```diff
@@ -0,0 +1,15 @@
+  tokenService,
+  userRepository,
+}) {
+  const samlId = tokenService.extractSamlId(externalUserToken);
```
Pourquoi catcher une exception dans le decodeToken , returner false en cas d'erreur et envoyer une autre exception Invalid Temporary key , pour la catcher et renvoyer une InvalidExternalUserTokenError.

Je propose de renommer juste la première exception d'origine et la renommer en InvalidToken et la laisser se propager.

chrc replied:
Faire le tour des effets de bord (api et fronts).

MathieuGilet replied:
Je le ferai plutôt dans une autre PR, celle-ci est déjà volumineuse et comme dit Christophe, il risque d'y avoir des impacts.

MathieuGilet commented on api/lib/domain/services/token-service.js:
```diff
@@ -50,6 +50,11 @@ function getDecodedToken(token) {
   }
 }
 
+function extractSamlId(token) {
```
Il y a déjà une méthode `extractExternalUserFromIdToken` qui extrait tout ce qui est contenu dans l'idToken mais si vous préférez avoir une méthode spécifique pour le `samlId`, ça me va aussi

octo-topi replied:
@MathieuGilet  
De cette manière, c'est le use-case qui remonte l'erreur `InvalidExternalUserTokenError` au lieu de `InvalidTemporaryKeyError`, les logs seront donc plus lisibles. 

Il y a une question à se poser sur la fonction intermédiaire (déjà appelée par un autre use-case) , qui perd la stacktrace, mais elle a déjà été identifiée par @yahyayouness et fera partie d'un ticket séparé
```
function getDecodedToken(token) {
  catch (err) {
    return false;
  }
```

MathieuGilet commented on api/lib/infrastructure/repositories/user-repository.js:
```diff
@@ -432,6 +433,19 @@ module.exports = {
+    return BookshelfUser
+      .where({ id: userId })
+      .save({ samlId }, { patch: true, method: 'update' })
+      .then(() => true)
```
Je ne pense que le `then(() => true)` soit utile ici.

chrc replied:
On ne veut pas retourner l'objet user. Le then() permet de retourner une autre valeur.

jonathanperret replied:
@chrc 

```js
    await BookshelfUser
      .where({ id: userId })
      .save({ samlId }, { patch: true, method: 'update' });

	return true;
```

MathieuGilet commented on api/tests/integration/infrastructure/repositories/user-repository_test.js:
```diff
@@ -1331,4 +1335,37 @@ describe('Integration | Infrastructure | Repository | UserRepository', () => {
+      const result = await userRepository.updateSamlId({ userId, samlId: expectedSamlId });
+
+      // then
+      expect(result).to.be.true;
```
Je comprends pourquoi on renvoie `true` du coup mais je me demande s'il ne vaudrait pas mieux de faire un `await knex('users').where({ id: userId })` et d'ensuite vérifier la valeur du `samlId`.

chrc replied:
Done

MathieuGilet commented on api/lib/application/users/index.js:
```diff
@@ -420,6 +420,49 @@ exports.register = async function(server) {
+          payload: Joi.object({
+            data: {
+              id: Joi.number().integer().positive().required(),
+              type: Joi.string().required(),
```
Plutôt que dire que le type est requis, spécifier directement sa valeur.

chrc replied:
Done

MathieuGilet commented:
J'ai voulu tester mes les données présente dans la description de la PR ne semblent plus valide. De toute façon il semble y avoir encore des changements en cours donc j'attends. Sinon, quelques petites choses ici et là qui pourraient être changées mais dans l'ensemble c'est très bien. Un grand bravo néanmoins pour le découpage des commits qui rend la review assez facile malgré le nombre de lignes. J'attends juste de tester fonctionnellement pour donner ma coche (et vérifier les derniers changements si il y en a).

chrc commented:
> J'ai voulu tester mes les données présente dans la description de la PR ne semblent plus valide. De toute façon il semble y avoir encore des changements en cours donc j'attends. Sinon, quelques petites choses ici et là qui pourraient être changées mais dans l'ensemble c'est très bien. Un grand bravo néanmoins pour le découpage des commits qui rend la review assez facile malgré le nombre de lignes. J'attends juste de tester fonctionnellement pour donner ma coche (et vérifier les derniers changements si il y en a).

Tu as testé en local ou sur la RA ?
@MathieuGilet : J'ai modifié le descriptif, car la date de naissance était erronée.

octo-topi commented on api/db/seeds/data/organizations-builder.js:
```diff
@@ -248,4 +248,21 @@ module.exports = function organizationsBuilder({ databaseBuilder }) {
     createdAt: new Date('2020-08-14')
   });
 
+  // PIX-1095
```
```suggestion
```

chrc replied:
Done

MathieuGilet reviewed (approved):
Tests ok (pensez à exécuter la requête suivante après votre test: `update users set "hasSeenAssessmentInstructions" = false, "samlId" = null  where id = 100063;`)

chrc commented:
Bug relevé par @yahyayouness :
<img width="1427" src="https://user-images.githubusercontent.com/88607/92924904-f6d9bf00-f439-11ea-9f9d-d7073f9ab1d9.png">

yahyayouness commented on api/lib/domain/errors.js:
```diff
@@ -320,6 +320,12 @@ class InvalidCertificationCandidate extends DomainError {
 }
 
+class InvalidExternalUserTokenError extends DomainError {
+  constructor(message = 'L\'idToken de l\'utilisateur externe est invalide.') {
```
On peut enlever d'échapper en mettant ‘
```suggestion
  constructor(message = 'L‘idToken de l‘utilisateur externe est invalide.') {
```

chrc replied:
Done

yahyayouness commented on api/lib/domain/errors.js:
```diff
@@ -320,6 +320,12 @@ class InvalidCertificationCandidate extends DomainError {
   }
 }
 
+class InvalidExternalUserTokenError extends DomainError {
```
Il faut la mapper dans l'error manager avec un unauthorized error

yahyayouness commented on mon-pix/app/components/routes/login-form.js:
```diff
@@ -1,50 +1,69 @@
-  @action
-  togglePasswordVisibility() {
-    this.toggleProperty('isPasswordVisible');
+    await this._authenticate(password, login);
```
En cas d'erreur authentification, on devrait pas ajouter une methode d'authentification. Il me semble que c'était fixé ?

chrc committed:
[APP] Add start campaign workflow acceptance tests

If the user is already reconciled and has no GAR authentication method yet,
add the GAR authentication method to this user.
